src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -81,28 +93,37 @@ namespace Quantum
             layerStartTime = lastLayer.startTime + lastLayer.duration;
         }
         this->metricsByLayer.emplace_back(
-            Layer {layerStartTime, max(this->preferredLayerDuration, minRequiredDuration)});
+            Layer{layerStartTime, max(this->preferredLayerDuration, minRequiredDuration)});
 
         return this->metricsByLayer.size() - 1;
     }
 
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::GetEffectiveBarrier
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::GetEffectiveBarrier() const
+    {
+        return CTracer::LaterLayer(this->globalBarrier, this->latestConditionalFence);
+    }
+
     //------------------------------------------------------------------------------------------------------------------
     // CTracer::FindLayerToInsertOperationInto
     //------------------------------------------------------------------------------------------------------------------
     LayerId CTracer::FindLayerToInsertOperationInto(Qubit q, Duration opDuration) const
     {
         const QubitState& qstate = this->UseQubit(q);
 
-        LayerId layerToInsertInto = INVALID;
+        LayerId layerToInsertInto = REQUESTNEW;
 
-        const LayerId firstLayerAfterBarrier =
-            this->globalBarrier == INVALID
-                ? this->metricsByLayer.empty() ? INVALID : 0
-                : this->globalBarrier + 1 == this->metricsByLayer.size() ? INVALID : this->globalBarrier + 1;
+        const LayerId barrier = this->GetEffectiveBarrier();
+        const LayerId firstLayerAfterBarrier = barrier == INVALID ? this->metricsByLayer.empty() ? REQUESTNEW : 0
+                                               : barrier + 1 == this->metricsByLayer.size() ? REQUESTNEW
+                                                                                            : barrier + 1;

```

Comment:
Minor:
I believe everything is correct here, but for those who are uncertain, this fragment can encourage them to spend time for double-checking the correctness of logic here. I would recommend placing parentheses explicitly.
Also [FYI](https://www.viva64.com/en/b/0391/#ID4F03F23BDD).